src/rate_limiter/src/lib.rs

```diff
@@ -471,6 +519,52 @@ impl Default for RateLimiter {
     }
 }
 
+/// State for saving a RateLimiter.
+#[derive(Versionize)]
+pub struct RateLimiterState {
+    ops: Option<TokenBucketState>,
+    bandwidth: Option<TokenBucketState>,
+    timer_active: bool,
+}
+
+impl Persist for RateLimiter {
+    type State = RateLimiterState;
+    type ConstructorArgs = ();
+    type Error = io::Error;
+
+    fn save(&self) -> Self::State {
+        RateLimiterState {
+            ops: self.ops.as_ref().map(|ops| ops.save()),
+            bandwidth: self.bandwidth.as_ref().map(|bw| bw.save()),
+            timer_active: self.timer_active,
```

Comment:
I don't think we need to save `timer_active`. We actually need to know how much of the timer is left so that the restore is also correct. If the timer was armed, instead of starting off at the remaining duration, we start the timer at  `TIMER_REFILL_STATE`.

I don't have enough knowledge about the rate limiter so I don't really know what is the appropriate state here.

One proposal would be to define a `reset_time_ms` u32 value instead with the following meaning:
- value = 0 -> timer is disarmed -> we just initialize the timer with the default state & set timer_active = false
- 0 < value <= REFILL_TIMER_INTERVAL_MS -> we initialize the timer and set the state to `TimerState::Oneshot(Duration::from_millis(value))` and set timer_active = true
- value > REFILL_TIMER_INTERVAL_MS -> the state is wrong and we should exit with an error.

Reply:
The rate-limiter budget is passively updated by its users (when they consume, they also update the budget based on elapsed time).
When there is insufficient budget to serve a request, the rate-limiter _blocks_, and users of the ratelimiter might _block_ as well (our devices in Firecracker do so). A one-shot timer is activated to _actively_ update the ratelimiter budget in case all its users are blocked and there is no one to _passively_ update it.

This timer has hardcoded period of 100ms. I think we can avoid the added complexity of serializing and deserializing remaining duration and just let it run 100ms after deserialization.

Reply:
I would actually go a step further and not even serialize whether the rate-limiter has pending timer.

I'm now thinking to just serialize budget-related state. On deserialization start off _unblocked_ (regardless of blocked/unblocked at serialization).

Post-deser all devices (all async objects really) need to be kicked anyway, so they will attempt to consume rate-limiter budget. If there is enough, it will go through, otherwise live ratelimiter code logic will do the blocking and timer setup.

This reduces ser/deser complexity, while being correct and fair behavior.

What do you think?

Reply:
Not serializing the `timer_active` at all and just not have the timer active sounds good to me.

Reply:
Removed `timer_active`, now the rate limiter starts with the timer turned off.